### PR TITLE
Make Ray compatible with `pydantic>=2.5.0`

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

This change makes Ray compatible with `pydantic>=2.5.0`. It makes all Ray unit tests use `pydantic==2.5.0`. It also adds a test suite to check that Ray Serve works with `pydantic<2`.

**Additional Changes:**

* This PR also updates the following packages:
  * `mypy`: `0.982` -> `1.7.0`
  * `deepspeed`: `0.8.3` -> `0.12.3`

**Follow-up:**
* We need to make a follow-up PR that makes Ray release tests run with `pydantic>=2.5.0`.

## Related issue number

Addresses:
* https://github.com/ray-project/ray/issues/38977
* https://github.com/ray-project/ray/issues/41095
* https://github.com/ray-project/ray/issues/39722
* https://github.com/ray-project/ray/issues/39864
* https://github.com/ray-project/ray/issues/39161
* https://github.com/ray-project/ray/issues/37019
* https://github.com/ray-project/ray/issues/37372
* https://github.com/ray-project/ray/pull/39162

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change updates Ray unit tests.

### CODE REVIEW

File: doc/requirements-doc.txt
Comment (can-anyscale) [inline doc/requirements-doc.txt:25]: you might need to follow the instructions (Updating dependencies and recompiling) in https://www.notion.so/anyscale-hq/OSS-Python-dependency-management-f32633b0018c423f927727807ea9da08

Maybe even just unpin the version and let the compiler find a suitable version

Comment (edoakes) [reply doc/requirements-doc.txt:25]: I'm attempting to follow those 😓

File: python/ray/tests/test_pydantic_serialization.py
Comment (shrekris-anyscale) [inline python/ray/tests/test_pydantic_serialization.py:158]: Could you clarify this `TODO`?

**Review by shrekris-anyscale — APPROVED** [review]

File: python/ray/tests/test_pydantic_serialization.py
Comment (edoakes) [reply python/ray/tests/test_pydantic_serialization.py:158]: still WIP i'll add a comment

**Review by aslonnie — COMMENTED** [review]
(seems @can-anyscale is going to review this one)

**Comment by edoakes** [comment]
FYI to reviewers: this is a WIP, I won't merge it until after the next `pydantic` release that includes our `cloudpickle` changes.

File: doc/requirements-doc.txt
Comment (ddelange) [reply doc/requirements-doc.txt:25]: > Maybe even just unpin the version and let the compiler find a suitable version

it's unpinned in setup.py

**Comment by ddelange** [comment]
> FYI to reviewers: this is a WIP, I won't merge it until after the next `pydantic` release that includes changes to make models `cloudpickle`-serializable.

do you have a link for this? will that mean that this PR will have to require `pydantic>=2.5`?

File: doc/requirements-doc.txt
Comment (edoakes) [reply doc/requirements-doc.txt:25]: ah yeah he was referring to in our internal build system for the pinned test requirements

**Comment by edoakes** [comment]
@ddelange here's the link to the Pydantic issue: https://github.com/pydantic/pydantic/issues/6763

After that is merged and released, we will require either `< 2.0` or `>= 2.5` (assuming `2.5` is the the release containing the fixes).

File: python/setup.py
Comment (pcmoritz) [inline python/setup.py:254]: Would it make sense to do `pydantic >= 2.5, < 2" here so we can support both? (given that the code seems to have already been written with that in mind and we seem to have tests for both)

That's also _roughly_ what fastAPI does: https://github.com/tiangolo/fastapi/blob/master/pyproject.toml#L44

Comment (shrekris-anyscale) [reply python/setup.py:254]: That makes sense, thanks for the catch! I updated the version.

File: python/ray/tests/test_client.py
Comment (ddelange) [inline python/ray/tests/test_client.py:14]: does this imply that the test suite will only run on pydantic v2?

if yes, does it make sense to drop pydantic v1?

Comment (ddelange) [reply python/ray/tests/test_client.py:14]: ah nvm, just spotted `EXPECTED_PYDANTIC_VERSION=1.10.12` in the test suite

Comment (shrekris-anyscale) [reply python/ray/tests/test_client.py:14]: We'll run all the Ray tests on Pydantic 2, and we'll also run the Serve tests with Pydantic 1.

File: python/ray/tests/test_pydantic_serialization.py
Comment (shrekris-anyscale) [reply python/ray/tests/test_pydantic_serialization.py]: I'll leave this as a `TODO` for now since it's not a regression.

**Review by can-anyscale — APPROVED** [review]
CI changes look good to me, thanks for working on this!

**Review by kouroshHakha — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]

File: python/requirements.txt
Comment (shrekris-anyscale) [reply python/requirements.txt]: Thanks for the suggestion! I added the upper bound. I'll omit the lower bound for now since we haven't done rigorous testing to decide on a good lower bound.

**Review by matthewdeng — APPROVED** [review]

**Review by aslonnie — APPROVED** [review]

File: python/requirements.txt
Comment (ericl) [inline python/requirements.txt:61]: >=2.5.0,<3 doesn't work?

**Review by ericl — APPROVED** [review]
Stamping for setup.py changes

File: python/requirements.txt
Comment (shrekris-anyscale) [reply python/requirements.txt:61]: That works, but we want to allow users to use `pydantic<2` for Serve, so no need to pin it to `2.5.0` explicitly for now.

Comment (ericl) [reply python/requirements.txt:61]: Can you add a comment to this effect?

Comment (shrekris-anyscale) [reply python/requirements.txt]: Yeah, I added one.

**Comment by shrekris-anyscale** [comment]
@architkulkarni This change is ready to merge.

**Comment by architkulkarni** [comment]
@shrekris-anyscale Looks like premerge is still running, please let me know again when it's ready

**Comment by shrekris-anyscale** [comment]
@architkulkarni Premerge passed. This PR is ready to merge.

**Comment by edoakes** [comment]
Happy to see this, thanks for pushing it forward @shrekris-anyscale

**Pull request merged**
